<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from scene.urdf.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="ur5_pick_place" xmlns:xacro="http://www.ros.org/wiki/xacro">
  <material name="white">
    <color rgba="1 1 1 1"/>
  </material>
  <material name="brown">
    <color rgba="0.5 0.25 0.01 1"/>
  </material>
  <material name="red">
    <color rgba="0.8 0 0 0.8"/>
  </material>
  <material name="green">
    <color rgba="0 0.8 0 0.8"/>
  </material>
  <!-- <xacro:include filename="$(find roadmap_tutorial)/urdf/ur.transmission.xacro" /> -->
  <!-- <xacro:include filename="$(find roadmap_tutorial)/urdf/ur.gazebo.xacro" /> -->
  <!-- measured from model -->
  <!--property name="shoulder_height" value="0.089159" /-->
  <!--property name="shoulder_offset" value="0.13585" /-->
  <!-- shoulder_offset - elbow_offset + wrist_1_length = 0.10915 -->
  <!--property name="upper_arm_length" value="0.42500" /-->
  <!--property name="elbow_offset" value="0.1197" /-->
  <!-- CAD measured -->
  <!--property name="forearm_length" value="0.39225" /-->
  <!--property name="wrist_1_length" value="0.093" /-->
  <!-- CAD measured -->
  <!--property name="wrist_2_length" value="0.09465" /-->
  <!-- In CAD this distance is 0.930, but in the spec it is 0.09465 -->
  <!--property name="wrist_3_length" value="0.0823" /-->
  <!-- manually measured -->
  <link name="base_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/base.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/base.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="4.0"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <inertia ixx="0.00443333156" ixy="0.0" ixz="0.0" iyy="0.00443333156" iyz="0.0" izz="0.0072"/>
    </inertial>
  </link>
  <joint name="shoulder_pan_joint" type="revolute">
    <parent link="base_link"/>
    <child link="shoulder_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.089159"/>
    <axis xyz="0 0 1"/>
    <limit effort="150.0" lower="-1.57" upper="1.57" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="shoulder_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/shoulder.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/shoulder.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="3.7"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <inertia ixx="0.010267495893" ixy="0.0" ixz="0.0" iyy="0.010267495893" iyz="0.0" izz="0.00666"/>
    </inertial>
  </link>
  <joint name="shoulder_lift_joint" type="revolute">
    <parent link="shoulder_link"/>
    <child link="upper_arm_link"/>
    <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.13585 0.0"/>
    <axis xyz="0 1 0"/>
    <limit effort="150.0" lower="-1.57" upper="0" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="upper_arm_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/upperarm.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/upperarm.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="8.393"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.28"/>
      <inertia ixx="0.22689067591" ixy="0.0" ixz="0.0" iyy="0.22689067591" iyz="0.0" izz="0.0151074"/>
    </inertial>
  </link>
  <joint name="elbow_joint" type="revolute">
    <parent link="upper_arm_link"/>
    <child link="forearm_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 -0.1197 0.425"/>
    <axis xyz="0 1 0"/>
    <limit effort="150.0" lower="0" upper="3.14" velocity="3.15"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="forearm_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/forearm.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/forearm.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="2.275"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.25"/>
      <inertia ixx="0.049443313556" ixy="0.0" ixz="0.0" iyy="0.049443313556" iyz="0.0" izz="0.004095"/>
    </inertial>
  </link>
  <joint name="wrist_1_joint" type="revolute">
    <parent link="forearm_link"/>
    <child link="wrist_1_link"/>
    <origin rpy="0.0 1.570796325 0.0" xyz="0.0 0.0 0.39225"/>
    <axis xyz="0 1 0"/>
    <limit effort="28.0" lower="-3.78" upper="1.22" velocity="3.2"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_1_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/wrist1.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/wrist1.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.219"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <inertia ixx="0.111172755531" ixy="0.0" ixz="0.0" iyy="0.111172755531" iyz="0.0" izz="0.21942"/>
    </inertial>
  </link>
  <joint name="wrist_2_joint" type="revolute">
    <parent link="wrist_1_link"/>
    <child link="wrist_2_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.093 0.0"/>
    <axis xyz="0 0 1"/>
    <limit effort="28.0" lower="-4.0" upper="1.0" velocity="3.2"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_2_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/wrist2.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/wrist2.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="1.219"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <inertia ixx="0.111172755531" ixy="0.0" ixz="0.0" iyy="0.111172755531" iyz="0.0" izz="0.21942"/>
    </inertial>
  </link>
  <joint name="wrist_3_joint" type="revolute">
    <parent link="wrist_2_link"/>
    <child link="wrist_3_link"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.09465"/>
    <axis xyz="0 1 0"/>
    <limit effort="28.0" lower="0" upper="0" velocity="3.2"/>
    <dynamics damping="0.0" friction="0.0"/>
  </joint>
  <link name="wrist_3_link">
    <visual>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/visual/wrist3.dae"/>
      </geometry>
      <material name="LightGrey">
        <color rgba="0.7 0.7 0.7 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/ur5/collision/wrist3.stl"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.1879"/>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <inertia ixx="0.0171364731454" ixy="0.0" ixz="0.0" iyy="0.0171364731454" iyz="0.0" izz="0.033822"/>
    </inertial>
  </link>
  <!-- <joint name="${prefix}ee_fixed_joint" type="fixed"> -->
  <!--   <parent link="${prefix}wrist_3_link" /> -->
  <!--   <child link = "${prefix}ee_link" /> -->
  <!--   <origin xyz="0.0 ${wrist_3_length} 0.0" rpy="0.0 0.0 ${pi/2.0}" /> -->
  <!-- </joint> -->
  <!-- <link name="${prefix}ee_link"> -->
  <!--   <collision> -->
  <!--     <geometry> -->
  <!--       <box size="0.01 0.01 0.01"/> -->
  <!--     </geometry> -->
  <!--     <origin rpy="0 0 0" xyz="-0.01 0 0"/> -->
  <!--   </collision> -->
  <!-- </link> -->
  <!-- <xacro:ur_arm_transmission prefix="${prefix}" /> -->
  <!-- <xacro:ur_arm_gazebo prefix="${prefix}" /> -->
  <!-- ROS base_link to UR 'Base' Coordinates transform -->
  <!-- <link name="${prefix}base"/> -->
  <!-- <joint name="${prefix}base_link-base_fixed_joint" type="fixed"> -->
  <!--   <origin xyz="0 0 0" rpy="0 0 ${-pi}"/> -->
  <!--   <parent link="${prefix}base_link"/> -->
  <!--   <child link="${prefix}base"/> -->
  <!-- </joint> -->
  <!-- Frame coincident with all-zeros TCP on UR controller -->
  <link name="tool0"/>
  <joint name="wrist_3_link-tool0_fixed_joint" type="fixed">
    <origin rpy="-1.570796325 0 0" xyz="0 0.0823 0"/>
    <parent link="wrist_3_link"/>
    <child link="tool0"/>
  </joint>
  <!-- Create the cart that is the robot's base -->
  <link name="cart">
    <visual>
      <origin xyz="0 0 0"/>
      <geometry>
        <box size="0.4572 0.7112 0.885825 "/>
      </geometry>
      <material name="brown"/>
    </visual>
    <collision>
      <origin xyz="0 0 0"/>
      <geometry>
        <box size="0.4572 0.7112 0.885825 "/>
      </geometry>
    </collision>
  </link>
  <!-- Attach the ur5 to the cart -->
  <joint name="cart_to_base_link" type="fixed">
    <origin rpy="0 0 0" xyz="0.1127125 0 0.4429125"/>
    <parent link="cart"/>
    <child link="base_link"/>
  </joint>
  <!-- Attach gripper -->
  <joint name="wrist_to_adapter_plate" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 -0.01"/>
    <parent link="tool0"/>
    <child link="adapter_plate"/>
  </joint>
  <link name="adapter_plate">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.007"/>
      <geometry>
        <cylinder length="0.014" radius="0.04"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.007"/>
      <geometry>
        <cylinder length="0.014" radius="0.04"/>
      </geometry>
    </collision>
  </link>
  <joint name="plate_to_gripper" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.014"/>
    <parent link="adapter_plate"/>
    <child link="robotiq_85"/>
  </joint>
  <link name="robotiq_85">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/gripper.stl"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/gripper.stl"/>
      </geometry>
    </collision>
  </link>
  <!-- Bounding volume, uncomment the below to enable -->
  <joint name="plate_to_bounding_volume" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.014"/>
    <parent link="adapter_plate"/>
    <child link="bounding_volume"/>
  </joint>
  <link name="bounding_volume">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0.0735"/>
      <geometry>
        <cylinder length="0.147" radius="0.0775"/>
      </geometry>
      <material name="red"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.0735"/>
      <geometry>
        <cylinder length="0.147" radius="0.0775"/>
      </geometry>
    </collision>
  </link>
  <link name="world"/>
  <!-- Attach the world to the robot chain -->
  <joint name="world_to_cart_base" type="fixed">
    <origin rpy="0 0 0" xyz="-0.2286 0 0.4429125"/>
    <parent link="world"/>
    <child link="cart"/>
  </joint>
  <!-- Attach world to the frame -->
  <joint name="world_to_frame" type="fixed">
    <origin rpy="0 0 1.57079632679" xyz="1.0 -1.25 0"/>
    <parent link="world"/>
    <child link="frame"/>
  </joint>
  <link name="frame">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/frame.dae"/>
      </geometry>
      <material name="black"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://roadmap_tutorial/meshes/frame.dae"/>
      </geometry>
      <material name="black"/>
    </collision>
  </link>
  <!-- Attach the world to the table -->
  <joint name="world_to_table" type="fixed">
    <origin rpy="0 0 0" xyz="0.5 0 0.92075"/>
    <parent link="world"/>
    <child link="table"/>
  </joint>
  <link name="table">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.015"/>
      <geometry>
        <box size="0.75 2.25 0.03"/>
      </geometry>
      <material name="white"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 -0.015"/>
      <geometry>
        <box size="0.75 2.25 0.03"/>
      </geometry>
      <material name="white"/>
    </collision>
  </link>
  <!-- Attach a box representing target object bounds -->
  <!-- Uncomment below block to visualize target bounds -->
  <!--    <joint name="base_link_to_target_bounds" type="fixed">  -->
  <!--   <origin rpy="0 0 0" xyz="0.45 0 ${0.275 - 0.147 - 0.014}" /> -->
  <!--   <parent link="base_link"/>  -->
  <!--   <child link="target_bounds"/>  -->
  <!-- </joint>  -->
  <!-- <link name="target_bounds">  -->
  <!--   <visual>  -->
  <!--     <origin rpy="0 0 0" xyz="0 0 0"/>  -->
  <!--     <geometry>  -->
  <!--       <box size="0.5 0.8 0.15"/>  -->
  <!--     </geometry>  -->
  <!--     <material name="green"/>  -->
  <!--   </visual> -->
  <!-- </link> -->
</robot>

